### ISSUE TITLE
Check the implementation of an overloaded function against all its overloaded signatures (or document the current behaviour)

### ISSUE THREAD

**Issue by jonathan-laurent** [microsoft/pyright#8248]
I was very surprised to see that neither Pyright nor Mypy typechecks the implementation of an overloaded function against all its overloaded signatures. For example, the bug in the function below is ignored silently:

```python
# No error nor warning is issued, using pyright 1.1.369 (strict mode)

from typing import assert_type, overload

@overload
def f(action: int) -> int: ...

@overload
def f(action: str) -> str: ...

def f(action: int | str) -> int | str:
    match action:
        case int():
            return 0
        case str():
            return 0  # Undetected bug: a string should be returned!

res = f("foo")
assert_type(res, str)
```

This behaviour is not documented in [PEP 484](https://peps.python.org/pep-0484/#function-method-overloading), in the [typing doc](https://typing.readthedocs.io/en/latest/spec/overload.html) or in the [Pyright doc](https://microsoft.github.io/pyright/#/typed-libraries?id=overloads). The [Mypy](https://mypy.readthedocs.io/en/stable/more_types.html#function-overloading) documentation does not mention such a check being made but does not exclude it explicitly either.

This behaviour strikes me as unintuitive and a major hole in the type system. Is there anywhere where it is being discussed? I can foresee some potential implementation difficulties (e.g. how should functions like `assert_type` and `reveal_type` behave within the implementation) but I do not see any insurmountable obstacle.

In any case, I would argue that the current behaviour must be clearly documented, with big warning boxes. Also, if this behaviour is intended, I would suggest adding a flag that disables overloaded definitions outside of stub files. Whether or not such a flag should be turned on by default in strict mode is debatable.

**Comment by erictraut** [comment]
Both pyright and mypy verify that the overload signatures are consistent with the signature of the implementation. The same is true for other type checkers and compilers in other languages that support overloads (like TypeScript). I'm not aware of any type checkers or compilers that type check the implementation against all signatures. In practice, that's not really feasible because the overloaded signatures do not need to be compatible with the implementation. For example, positional-only parameter names may legally differ between an overload signature and an implementation signature. Even if this were possible to overcome, reporting errors within the implementation based on multiple validation passes would be very confusing because some errors would occur multiple times; others would occur only for specific overload signatures. Plus, the evaluated types of local variables within the implementation would differ from one evaluation to the next, which is very problematic for type evaluation engines designed for language servers (like pyright). For all of these reasons, it's unlikely we'd ever implement such a feature.
